File: GuardianMonsters/guardians/src/de/limbusdev/guardianmonsters/guardians/abilities/Node.java
package de.limbusdev.guardianmonsters.guardians.abilities;

import com.badlogic.ashley.signals.Signal;

/**
 * Node
 *
 * @author Georg Eckert 2017
 */

public class Node extends Signal<Node> {

    public enum Type {
        EMPTY, ABILITY, EQUIPMENT, METAMORPHOSIS,
    }

    public enum State {
        DISABLED, ENABLED, ACTIVE,
    }

    public int x;
    public int y;
    public int ID;
    public Type type;
    private State state;

    public Node(int x, int y, int ID) {
        this.y = y;
        this.x = x;
        this.ID = ID;
        this.type = Type.EMPTY;
        this.state = State.DISABLED;
    }

    public Node(int x, int y, int ID, Type type, State state) {
        this.x = x;
        this.y = y;
        this.ID = ID;
        this.type = type;
        this.state = state;
    }

    public void activate() {

        state = State.ACTIVE;
        dispatch(this);

    }

    public void enable() {

        if(state != State.ACTIVE) {
            state = State.ENABLED;
        }
        dispatch(this);
    }

    public void disable() {
        if(state != State.ACTIVE && state != State.ENABLED) {
            state = State.DISABLED;
        }
        dispatch(this);
    }

    public boolean isActive() {
        return (state == State.ACTIVE);
    }

    public boolean isEnabled() {
        return (state == State.ENABLED);
    }

    public boolean isDisabled() {
        return (state == State.DISABLED);
    }

    public State getState() {
        return state;
    }

    @Override
    public String toString() {
        return "Node " + Integer.toString(ID) + ": " + type.toString() + " & " + state.toString();
    }
}

File: GuardianMonsters/guardians/src/de/limbusdev/guardianmonsters/guardians/abilities/Node.kt
package de.limbusdev.guardianmonsters.guardians.abilities

import com.badlogic.ashley.signals.Signal

/**
 * Node
 *
 * A *Node* is one of the base elements of the @see AbilityGraph. It represents some sort of ability
 * a @see AGuardian can learn. It may also be in a specific state.
 *
 * @author Georg Eckert 2019
 */

class Node : Signal<Node>
{
    // ............................................................................... Inner Classes
    enum class Type  { EMPTY, ABILITY, EQUIPMENT, METAMORPHOSIS }

    enum class State { DISABLED, ENABLED, ACTIVE }

    // .................................................................................. Properties
    var x: Int = 0
    var y: Int = 0
    var ID: Int = 0
    var type:  Type
    var state: State

    // ................................................................................ Constructors
    init
    {
        type = Type.EMPTY
        state = State.DISABLED
    }

    constructor(x: Int, y: Int, ID: Int)
    {
        this.x = x
        this.y = y
        this.ID = ID
    }

    constructor(x: Int, y: Int, ID: Int, type: Type, state: State) : this(x, y, ID)
    {
        this.type = type
        this.state = state
    }

    // ........................................................................... Getters & Setters
    val isActive:   Boolean get() = state == State.ACTIVE;
    val isEnabled:  Boolean get() = state == State.ENABLED
    val isDisabled: Boolean get() = state == State.DISABLED


    fun activate()
    {
        state = State.ACTIVE
        dispatch(this)
    }

    fun enable()
    {
        if (state != State.ACTIVE) { state = State.ENABLED }
        dispatch(this)
    }

    fun disable()
    {
        if (state != State.ACTIVE && state != State.ENABLED) { state = State.DISABLED }
        dispatch(this)
    }

    override fun toString(): String = "Node $ID: $type & $state";
}
